modules/runner/src/integrationTest/java/org/apache/ignite/internal/sql/engine/ItDataTypesTest.java

```diff
@@ -86,6 +87,23 @@ public void testUnicodeStrings() {
         }
     }
 
+    /** Test correctness of char types limitation. */
+    @Test
+    public void charLimitation() {
+        sql("create table limitedChar (pk int primary key, f1 VARCHAR(5), f2 CHAR(4))");
+        assertThrows(CalciteContextException.class, () -> sql("insert into limitedChar(pk, f2) values (1, 'aaaadddd')"));
+        assertThrows(CalciteContextException.class, () -> sql("insert into limitedChar values (2, 'aaaaabbbbb', 'aaaadddd')"));
+        assertThrows(CalciteContextException.class, () -> sql("insert into limitedChar values (3, '我叫什么名字', '我叫什么名字')"));
+        sql("insert into limitedChar values (4, ' aaa    ', ' bbb     ')");
+
+        List<List<Object>> res = sql("select f1, f2 from limitedChar ORDER BY pk");
+
+        List<Object> r = res.get(0);
+
+        assertEquals(" aaa ", r.get(0));
```

Comment:
Insert '__aaa___'. Retrieve '_aaa_'. Do we still expect data truncation to occur?

Reply:
We also need to check the case, when an input is a dynamic parameter and a target column's type has precision that is less then a value of the dynamic parameter.

Reply:
> Insert '...aaa..... Retrieve '.aaa.'. Do we still expect data truncation to occur?
It seems that removing trailing blank characters is a correct behaviour.